rfd/001-tcp-load-balancer-design-doc.md

```diff
@@ -0,0 +1,119 @@
+
+Generate private (client.crt) and public (client.key) keys for the client:
+$ openssl genrsa -out client.key
+$ openssl req -new -key client.key -out client_reqout.txt
+$ openssl x509 -req -in client_reqout.txt -days 3650 -sha256 -CAcreateserial -CA clentca.cert -CAkey clientcakey.pem -out client.crt		
+```
+
+For this project, a test client will be created and provided with a valid client cert for testing purposes. These initial dummy certificates and keys will be included (as plaintext) within this repository for the the benefit of reviewers, but should stored as secure secrets and managed outside of this repo in a future version.
+
+Additionally, the requirement of over-the-wire encryption is not mentioned, so even though the authentication requires certs and keys, the data sent after authentication may or may not be encrypted depending on what is most convenient at the point of implementation.
+
+### Rate Limiter
+This load balancer will implement a per-client connection rate limiter that tracks the number of client connections, and limits to n (configurable) active connections per client. It could track client identity based on attributes like IP address, and device identifier, but since this load balancer requires authentication (mTLS), we will identify unique clients based on a V5 UUID generated from the client's TLS certificate. 
+
+```
+clientID := uuid.NewV5(loadBalancerID, client.TLSCertificate().String())
+```
+
+Client connection counts can be stored in a `sync.Map` structure for quick modification, and safety across concurrent goroutines. Regarding this map choice, the [`sync` package documentation](https://pkg.go.dev/sync#Map) states the following: 
+> "The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex."
+
+Our use-case follows the #2 reason for leveraging `sync.Map` over a standard map with a Mutex; our disjoint sets of keys will be the unique client ids.
+
+
+### Upstream Authorization Management
+
+The load balancer will employ a simple authorization scheme which will deny access to specific upstreams for certain clients. This scheme will be statically defined in code. Once the load balancer has determined that the client is authenticated and within client connection limit (and prior to selecting the least connections host) it will filter out unauthorized hosts for that client. To simply demonstrate this capability, we store a map of `clientID`s (based on client certificate) to disallowed `hostID`s; because host configuration is outside the scope of this project, the order in which the host was registered with the load balancer will be used as the `hostID`.
```

Comment:
Yes, good callout. I'll add more details in this part of the design. 

The plan mentioned in the rate-limiting section alludes to generating and persisting a V5 UUID during the TLS handshake, when the client cert is presented to the server for validation. However, leveraging the standard `crypto/tls` library (which is what I think will be used) may make it cumbersome to access that cert during the handshake. It's probably going to be easier to have the client expose the cert, and then access it from the LB when connection is established. 

Since we're hard-coding clients for the scope of this project, it should be simple to expose the cert on the client struct (unless I come across a more elegant method during implementation).

```
clientID := uuid.NewV5(loadBalancerID, client.TLSCertificate().String())
```

What are your thoughts on that approach?

Reply:
> by some certificate value instead of whole cert hash?
Is there a particular value you would recommend?

Perhaps the common name (FQDN) could be a better identifier since it is more likely to remain constant through cert expiration/renewal, but it also has the common flaws of a client identifier (like not being permanent, and potential for duplicates).

Additionally, it appears that the certificate hash is the standard within the MTLS Sender Constrained Tokens specification. See [tools.ietf draft of sender constrained access tokens](https://tools.ietf.org/id/draft-ietf-oauth-mtls-07.html#SenderConstrainedAccess) and [curity article on sender constrained access tokens](https://curity.io/resources/learn/oauth-certificate-bound-access-token/#mutual-tls-sender-constrained-tokens) for more details. Which would incline me towards sticking with the cert hash.

Clients are certainly subject to change, so if we needed to build a robust per-client authorization system, I would probably opt for an authorization server to return sender constrained tokens, which clients can obtain through a registration processes prior to accessing the load balancer.

Reply:
Additionally, it appears that a hash of the certificate is standard for Sender Constrained Tokens. See [tools.ietf draft of sender constrained access tokens](https://tools.ietf.org/id/draft-ietf-oauth-mtls-07.html#SenderConstrainedAccess) or [curity article on sender constrained access tokens](https://curity.io/resources/learn/oauth-certificate-bound-access-token/#mutual-tls-sender-constrained-tokens) for references to that cert being hashed as the thumbprint to be included with tokens. This would incline me towards sticking with the cert hash.

Reply:
@rosstimothy and @smallinsky you both make good points. The certificate constrained access tokens are focused on ensuring the user is authorized only from the same client that it used to authenticate. 

My proposal to apply the same logic to clientIDs does not make sense. I mixed up clients and users when thinking through this.

Going back to @smallinsky 's question:
> What would you say about distinguishing user by some certificate value instead of whole cert hash ?

What do you think of using the FQDN for generating the clientID? It would still require some updates when domain names change, but should be more permanent than the certificate. Also when there are duplicates, they should be representing the same client.

Reply:
Sounds good. Thanks for helping us come to a conclusion on this, and for teaching me some things in the process!

Reply:
I've updated the section on clientIDs to reference the FQDN for identifying clients.